### PR TITLE
fix(weave): map openai cached_tokens to cache_read on otel ingest

### PR DESCRIPTION
**Pull request by gtarpenning**
## Summary
- OTel agent-span ingest only recognized the Anthropic-style `gen_ai.usage.cache_read.input_tokens` key, so OpenAI-compatible / LiteLLM-bridged harnesses reporting cached tokens under `prompt_tokens_details`/`input_tokens_details` had them silently dropped, overstating cost
- Adds the OpenAI-shaped keys as fallback lookups in both the agent-span extraction path (`semconv.py`) and the raw-OTLP calls path (`constants.py`), with the canonical Anthropic key kept highest priority
- Jira: [WB-37554](https://coreweave.atlassian.net/browse/WB-37554)

## Testing
unit test: openai-shaped cached_tokens (both prompt_tokens_details and input_tokens_details shapes) now populate cache_read_input_tokens; anthropic key still preferred when both present

## QA validation (zoo-qa)

| Master `89e03dc` | Candidate `8596b92` |
| --- | --- |
| master: Cache read 0 | candidate: Cache read 640 |

Identical live Agent-OTLP `chat` spans, with `gen_ai.usage.prompt_tokens_details.cached_tokens=640`, input `1000`, output `50`, and provider `openai`.

| build | persisted `cache_read_input_tokens` | evidence |
| --- | ---: | --- |
| master `89e03dc` | 0 | [Weave trace](https://app.qa.wandb.ai/weave-team/weave-codex-cache-read-7590-zoo-qa-master-20260717-104520/weave/agents?dashConv=zoo-qa-master-openai-cache-probe-7590-20260717-104520&convTurn=1&chatTrace=01c9dbcb9946b487da284085224bdec2) |
| candidate `8596b92` | 640 | [Weave trace](https://app.qa.wandb.ai/weave-team/weave-codex-cache-read-7590-zoo-qa-candidate-20260717-105120/weave/agents?dashConv=zoo-qa-candidate-openai-cache-probe-7590-20260717-105120&convTurn=1&chatTrace=edb15436cfb1d6f8550066ee4af0d7a0) |

### Exact Agent-OTLP logging probe

This inline logger was run unchanged against master and candidate. The runner supplied only `WANDB_BASE_URL`, `WANDB_API_KEY`, `PROBE_PROJECT`, and `PROBE_CONVERSATION` as environment variables; no credentials are embedded here.

```bash
node --input-type=module -e '
  import {context, SpanKind, trace} from "@opentelemetry/api";
  import {buildTracer} from "./node_modules/weave-codex/dist/spans/tracer.js";

  const config = {
    baseUrl: process.env.WANDB_BASE_URL,
    apiKey: process.env.WANDB_API_KEY,
    projectId: process.env.PROBE_PROJECT,
    entity: process.env.PROBE_PROJECT.split("/")[0],
    project: process.env.PROBE_PROJECT.split("/")[1],
    captureContent: false,
    debug: false,
  };
  const {tracer, shutdown} = buildTracer(config);
  const now = Date.now();

  const root = tracer.startSpan("invoke_agent codex", {
    kind: SpanKind.INTERNAL,
    startTime: new Date(now),
    attributes: {
      "gen_ai.operation.name": "invoke_agent",
      "gen_ai.provider.name": "openai",
      "gen_ai.conversation.id": process.env.PROBE_CONVERSATION,
      "gen_ai.agent.name": "codex",
      "gen_ai.request.model": "zai-org/GLM-5.2",
    },
  });

  const parent = trace.setSpan(context.active(), root);
  const chat = tracer.startSpan("chat zai-org/GLM-5.2", {
    kind: SpanKind.CLIENT,
    startTime: new Date(now + 1),
    attributes: {
      "gen_ai.operation.name": "chat",
      "gen_ai.provider.name": "openai",
      "gen_ai.conversation.id": process.env.PROBE_CONVERSATION,
      "gen_ai.request.model": "zai-org/GLM-5.2",
      "gen_ai.response.model": "zai-org/GLM-5.2",
      "gen_ai.usage.input_tokens": 1000,
      "gen_ai.usage.output_tokens": 50,
      "gen_ai.usage.prompt_tokens_details.cached_tokens": 640,
    },
  });

  chat.end(new Date(now + 2));
  root.end(new Date(now + 3));
  await shutdown();
  console.log(JSON.stringify({
    project: process.env.PROBE_PROJECT,
    conversation_id: process.env.PROBE_CONVERSATION,
    trace_id: root.spanContext().traceId,
    source_cached_tokens: 640,
  }));
'
```

[WB-37554]: https://coreweave.atlassian.net/browse/WB-37554?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: weave/trace_server/opentelemetry/constants.py [hunk 1]
```diff
@@ -112,6 +112,9 @@
     ],
     "cache_read_input_tokens": [
         ("gen_ai.usage.cache_read.input_tokens", try_parse_int),
+        ("gen_ai.usage.prompt_tokens_details.cached_tokens", try_parse_int),
+        ("gen_ai.usage.input_tokens_details.cached_tokens", try_parse_int),
+        ("prompt_tokens_details.cached_tokens", try_parse_int),
```
Comment (jtschoonhoven) [inline weave/trace_server/opentelemetry/constants.py:117]: TL;DR: if `input_tokens = 0` then this will fall back to `cached_tokens` which is probably not what you want.

The raw OTLP parser checks `if value`: before accepting each candidate ([context](https://github.com/wandb/weave/blob/d8b15277c68f6793909e04b8cb9594856ac4e621/weave/trace_server/opentelemetry/attributes.py#L64-L65)). Consequently, attributes containing canonical `gen_ai.usage.cache_read.input_tokens = 0` and fallback `gen_ai.usage.prompt_tokens_details.cached_tokens = 640` produce `cache_read_input_tokens = 640`. That contradicts the canonical-first contract tested by this PR and reports cache reads that the canonical attribute explicitly says did not occur. The new precedence test exercises `extract_genai_span`, whose lookup correctly uses the first non-None value, but it does not exercise get_weave_usage/USAGE_KEYS.



Recommended action: make this raw lookup distinguish a missing value from a valid zero (scoped so unrelated empty-value behavior is not unintentionally changed), and add a get_weave_usage or protobuf-to-call regression test with canonical zero plus a nonzero fallback that asserts zero wins.

**Review by ro31337 — APPROVED** [review]

File: weave/trace_server/opentelemetry/constants.py (hunk 1 above)
Comment (gtarpenning) [reply weave/trace_server/opentelemetry/constants.py:117]: yeah easy ty
